stempy/image.cpp

```diff
@@ -269,6 +270,62 @@ vector<STEMImage> createSTEMImages(InputIt first, InputIt last,
   return images;
 }
 
+// returns the binding result
+std::vector<double> getContainer(const STEMImage& inImage, const int numBins)
+{
+  // information about input STEMImage
+  int width = inImage.width;
+  int height = inImage.height;
+  auto curData = inImage.data;
+  std::cout << "Input STEM Image has width = " << width
+            << ", height = " << height << std::endl;
+
+  auto result = std::minmax_element(curData.get(), curData.get() + width * height);
+  double min = *result.first;
+  double max = *result.second;
+
+  // the "length" of each slot of the container
+  double length = static_cast<double>((max - min) / numBins);
```

Comment:
The way that you have implemented this, you are using the bins as a lower and upper bound on the value to see if it should be added to the bin. In that case in order to properly assign upper and lower bounds you should use a length of `(max - min) / (numBins - 1)`. That would avoid the necessity in the following function of adding max to the last slot - they should all be in range.

Example, I want four bins between 0 and 1, that would result in 0.0, 0.25, 0.50 and 0.75 using the logic above. If I used `numBins - 1` it would yield 0.0, 0.33, 0.66 and 1.0 fully covering the intended range.

Reply:
As max and min are already double I don't think the cast is needed either.

Reply:
Continue with your example, after adding max to my current method, returned `container` would include 4 intervals/bins; [0.0, 0.25), [0.25, 0.5), [0.5, 0.75) and [0.75, 1.0], which matches with user input `numBins=4`. 
The method you suggested does cover the intended range, but only generate 3 bins, which are [0.0, 0.33), [0.33, 0.66) and [0.66, 1.0). Did I miss anything here?

Reply:
I was unsure if `numBins` being an `int` would cause any rounding errors, but if are certain about this, I can definitely remove the extra cast. Thanks!

Reply:
Your code also looks like it needs `bins + 1`, how is it not walking off the end of the array? See the conditional below, using `value >= bins[i] && value < bins[i + 1]` where `i < numBins`. No rounding error would occur, if you want to force double precision (not needed here) you would need to cast `numBins`.

Reply:
When generating the container, my for loop is `for (int i = 0; i **<=** numBins; i++)`, so continue with your example, it would be 5 numbers; 0, 0.25, 0.5, 0.75 and 1 after the loop. 
However, when generating frequencies, the for loop is `for (int i = 0; i **<** numBins; i++)`, which is why it is not walking off the end of the array.